sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -1280,6 +1280,16 @@ case class Cast(
     }
   }
 
+  // Whether Spark SQL can evaluation the try_cast as the legacy cast, so that no `try...catch`
+  // is needed and the performance can be faster.
+  private lazy val canUseLegacyCastForTryCast: Boolean = {
+    (child.dataType, dataType) match {
+      case (StringType, _: FractionalType) => true
+      case (StringType, _: DatetimeType) => true
+      case _ => false
```

Comment:
Using Cast.canUpCast is not safe here...Decimal(19,0) can be cast as long since the decimal for long is Decimal(20, 0), while the max long value has only 19 digits. 
I will revisit this part later. 
On the other hand, if the cast can up cast to another type, adding the `try...catch` won't significantly affect the perf.